Audit/Computer/Distribution Groups Management Events - ECS  related.user field mapping

janniten opened the pull request:
Added Audit and Log Management related events, Computer Object Management Events, Distribution Groups Events - Change in user.name for user management events and related.user mapping

Event|Description
--|--
|1100|The event logging service has shut down.
|1102|The audit log was cleared.
|1104|The security log is now full.|
|1105|Event log automatic backup.|
|1108|The event logging service encountered an error while processing an incoming event published from %1
|4719|System audit policy was changed.|
|4741|A computer account was created.|
|4742|A computer account was changed.|
|4743|A computer account was deleted.|
|4744|A security-disabled local group was created.|
|4745|A security-disabled local group was changed.|
|4746|A member was added to a security-disabled local group.|
|4747|A member was removed from a security-disabled local group.|
|4748|A security-disabled local group was deleted.|
|4749|A security-disabled global group was created.|
|4750|A security-disabled global group was changed.|
|4751|A member was added to a security-disabled global group.|
|4752|A member was removed from a security-disabled global group.|
|4753|A security-disabled global group was deleted.|
|4759|A security-disabled universal group was created.|
|4760|A security-disabled universal group was changed.|
|4761|A member was added to a security-disabled universal group.|
|4762|A member was removed from a security-disabled universal group.|
|4763|A security-disabled global group was deleted.|


**New Events**
Due to that Windows events are the source of information for Winlogbeat the events 1100,1102,1104,1105,1108 and 4719 has been added in order to monitor changes in the audit policy configuration, log deletion and other failures in the log subsystem.
For event 4719 a human readable description was added in order to know which setting was modified (winlog.event_data.SubCategory) and to which value (winlog.event_data.AuditPolicyChangesDescription)


Distribution Groups (Security-Disabled) Management Events were added. Those events are processed in the same way and with the same function that Security Groups (https://github.com/elastic/beats/pull/14299).  In order to add information about the nature of the group being managed the *type* (Security-Disabled/Security-Enabled) and *scope* (Local,Global,Universal) where added as winlog.group.type and winlog.group.scope

ComputerObject Management events were added

**Changes to ECS mappings**

In https://github.com/elastic/ecs/issues/678 and https://github.com/elastic/ecs/issues/589 we have been discussing how n-ary relationship between users in an event should be named and mapping into ECS.
In https://github.com/elastic/beats/pull/13530 winlog.event_data.TargetUserName has been mapped to user.name but from the reasons exposed in https://github.com/elastic/ecs/issues/678 and https://github.com/elastic/ecs/issues/589 the mapping winlog.event_data.SubjectUserName -> user.name is more appropriate. This mapping was changed.

Also, with the adding of related fields in ECS 1.3 and specifically the related.user field (https://github.com/elastic/ecs/pull/694) all the user names appearing in one event were mapped to the related user events.
Every time a SubjectUserName or TargetUserName is copied also is added to the related.user field, as well as other users appearing in the event.

Event test data were added for all events with the exception of event 1108 which I was not able to reproduce

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4742": "changed-computer-account",
+        "4743": "deleted-computer-account",
+        "4744": "added-group-account",
+        "4745": "changed-group-account",
```
"changed-distribution-group-account"?

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4743": "deleted-computer-account",
+        "4744": "added-group-account",
+        "4745": "changed-group-account",
+        "4746": "added-group-account-to",
```
"added-member-to-distribution-group"?

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4744": "added-group-account",
+        "4745": "changed-group-account",
+        "4746": "added-group-account-to",
+        "4747": "deleted-group-account-from",
```
"deleted-member-from-distribution-group"?

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4745": "changed-group-account",
+        "4746": "added-group-account-to",
+        "4747": "deleted-group-account-from",
+        "4748": "deleted-group-account",
```
"deleted-distribution-group-account"?

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4746": "added-group-account-to",
+        "4747": "deleted-group-account-from",
+        "4748": "deleted-group-account",
+        "4749": "added-group-account",
```
"added-distribution-group-account"

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4747": "deleted-group-account-from",
+        "4748": "deleted-group-account",
+        "4749": "added-group-account",
+        "4750": "changed-group-account",
```
"changed-distribution-group-account"?

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4748": "deleted-group-account",
+        "4749": "added-group-account",
+        "4750": "changed-group-account",
+        "4751": "added-group-account-to",
```
"added-member-to-distribution-group"?

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4749": "added-group-account",
+        "4750": "changed-group-account",
+        "4751": "added-group-account-to",
+        "4752": "deleted-group-account-from",
```
"deleted-member-from-distribution-group"?

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4750": "changed-group-account",
+        "4751": "added-group-account-to",
+        "4752": "deleted-group-account-from",
+        "4753": "deleted-group-account",
```
"deleted-distribution-group"?

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
         "4756": "added-group-account-to",
         "4757": "deleted-group-account-from",
         "4758": "deleted-group-account",
+        "4759": "added-group-account",
```
"added-distribution-group-account"?

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
         "4757": "deleted-group-account-from",
         "4758": "deleted-group-account",
+        "4759": "added-group-account",
+        "4760": "changed-group-account",
```
"changed-distribution-group-account"

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
         "4758": "deleted-group-account",
+        "4759": "added-group-account",
+        "4760": "changed-group-account",
+        "4761": "added-group-account-to",
```
"added-member-to-distribution-group"?

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4759": "added-group-account",
+        "4760": "changed-group-account",
+        "4761": "added-group-account-to",
+        "4762": "deleted-group-account-from",
```
"removed-member-from-distribution-group"?

janniten replied:
ok!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4760": "changed-group-account",
+        "4761": "added-group-account-to",
+        "4762": "deleted-group-account-from",
+        "4763": "deleted-group-account",
```
"deleted-distribution-group-account"?

janniten replied:
ok!

janniten commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -46,14 +46,21 @@ var security = (function () {
+    // event.action Description Table
     // event.action Description Table
     var eventActionTypes = {
+        "1100": "logging-shutdown",
```
ok! I'll change it

janniten commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -72,18 +79,136 @@ var security = (function () {
+        "4741": "added-computer-account",
+        "4742": "changed-computer-account",
+        "4743": "deleted-computer-account",
+        "4744": "added-group-account",
```
ok!

janniten commented:
> Thanks this is great.
> 
> I just had some thoughts about alternate names, let me know what you think.
> 
> Would you like me to generate the golden files for you?

Hi @leehinman ,
I have updated the action description with your suggestions. Algo I have changed the action description for normal groups in order to be consistent with your proposal.
Regarding to golden files, yes, please Can you generate the files for me?

Thank you!

leehinman commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -1316,6 +1504,24 @@ var security = (function () {
+        if (!code) {
+         return;
+        }
+        evt.Put("winlog.group.type", group_types[code][0]);
```
Getting an error here for events 4764 and 4799, looks like those events don't have an entry in group_types.  We should probably add those and we might want to add a check to make sure there are entries before we do the evt.Put

leehinman commented:
> > Found a bug while generating the golden files.
> 
> Hi @leehinman . I have removed the Group Type related fields and functions. Now it should work when generating golden files.
> I think it does not add valuable information at the moment. I'll keep thinking in how to add that information for a future PR

@janniten Thanks.  I was able to produce the golden files.  I noticed one odd thing.  In security-windows2016_4723_Password_Change.evtx the Subject & Target user are the same, so when they get added to the related.user list we get a duplicates.  That seems like a bug to me.  What do you think?  I was thinking we could use a Set in addRelatedUser to make sure the values are unique.  Sound OK to you?

```
+    "related": {
+      "user": [
+        "Administrator",
+        "Administrator"
+      ]
```

One other small thing.  I noticed that 
```
"process": {
      "name": "null"
}
```

Was in a few of the golden files.  I think if we add

```
            if (exe) {
                evt.Put("process.name", path.basename(exe));
            }
```

to renameCommonAuthFields that will prevent the nulls

janniten commented:
Hi @leehinman
I have fixed the duplicates issues in the related.users array. I was not able to use Sets due that they are available in ES6 (as far as I know we are using ES 5.1) . I use a prototype and  anonymous function in order to define an "unique" function that returns the distinct elements of an array
If there is a better way to solve this, please let me know and I'll investigate.
I have also added the check for the null process in the renameCommonAuthFields function, but I use the following code in order to be consistent with the same check in other parts of the code.
```
 if (!exe) {
                return;
            }
 evt.Put("process.name", path.basename(exe));
```

How are the golden files generated? Next time I can try to generate the golden files for myself and detect errors in an early stage.
Thank you!

andrewkroh commented:
If the problem is that you are getting duplicate values in a string array, then you can use `event.AppendTo("related.user", "andrewkroh")`. It will not add a duplicate value. See the documentation in https://www.elastic.co/guide/en/beats/winlogbeat/7.5/processor-script.html.

janniten commented:
> If the problem is that you are getting duplicate values in a string array, then you can use `event.AppendTo("related.user", "andrewkroh")`. It will not add a duplicate value. See the documentation in https://www.elastic.co/guide/en/beats/winlogbeat/7.5/processor-script.html.

Thank you @andrewkroh.
I have already modified the code in order to use the script processor AppendTo.
![image](https://user-images.githubusercontent.com/33020901/71838188-9202e380-30b8-11ea-99b9-c8199d4501a6.png)

leehinman commented:
> Hi @leehinman
> I have fixed the duplicates issues in the related.users array. I was not able to use Sets due that they are available in ES6 (as far as I know we are using ES 5.1) . I use a prototype and anonymous function in order to define an "unique" function that returns the distinct elements of an array
> If there is a better way to solve this, please let me know and I'll investigate.
> I have also added the check for the null process in the renameCommonAuthFields function, but I use the following code in order to be consistent with the same check in other parts of the code.
> 
> ```
>  if (!exe) {
>                 return;
>             }
>  evt.Put("process.name", path.basename(exe));
> ```
> 
> How are the golden files generated? Next time I can try to generate the golden files for myself and detect errors in an early stage.
> Thank you!

Sorry about the Set suggestion, luckily @andrewkroh had a much better suggestion :-)

You can make the golden files by being in the "C:\Gopath\src\github.com\elastic\beats\x-pack\winlogbeat\module\security\test" directory and running "go test -update".

I found one more issue, in win2019 the SubcategoryGuid is lowercase, not uppercase so the addAuditInfo function wasn't adding subcategor and AuditPolicyChangesDescription.  Easy fix with adding toUpperCase before doing the lookup.  I'll add that to the same commit as the golden files.

janniten commented:
> > Hi @leehinman
> > I have fixed the duplicates issues in the related.users array. I was not able to use Sets due that they are available in ES6 (as far as I know we are using ES 5.1) . I use a prototype and anonymous function in order to define an "unique" function that returns the distinct elements of an array
> > If there is a better way to solve this, please let me know and I'll investigate.
> > I have also added the check for the null process in the renameCommonAuthFields function, but I use the following code in order to be consistent with the same check in other parts of the code.
> > ```
> >  if (!exe) {
> >                 return;
> >             }
> >  evt.Put("process.name", path.basename(exe));
> > ```
> > 
> > 
> > How are the golden files generated? Next time I can try to generate the golden files for myself and detect errors in an early stage.
> > Thank you!
> 
> Sorry about the Set suggestion, luckily @andrewkroh had a much better suggestion :-)
> 
> You can make the golden files by being in the "C:\Gopath\src\github.com\elastic\beats\x-pack\winlogbeat\module\security\test" directory and running "go test -update".
> 
> I found one more issue, in win2019 the SubcategoryGuid is lowercase, not uppercase so the addAuditInfo function wasn't adding subcategor and AuditPolicyChangesDescription. Easy fix with adding toUpperCase before doing the lookup. I'll add that to the same commit as the golden files.

Thank you @leehinman 
Is it ready to merge or shall I do something more?

neu5ron commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -1316,6 +1475,15 @@ var security = (function () {
+        if (!computer) {
+            return;
+    }
+        evt.Put("winlog.computer.name", computer.split(".")[0]);
```
I would not make this assumption (and the next line) that these are the computer and domain by splitting `.` - because many organizations can/do have multi level domains (ie: addomain.mydomain.local)

janniten replied:
@neu5ron Does it make sense this?

evt.Put("winlog.computer.name", computer.split(".")[0]);
evt.Put("winlog.computer.domain", computer.split(".").slice(1).join('.');

If computer name is computer.addomain.mydomain.local

winlog.computer.name -> computer
winlog.computer.domain -> addomain.mydomain.local

neu5ron replied:
Many of these windows logs will already have the domain, and for the one's that don't - if the log does not have the domain, unless one can explicitly know with certainty what the domain is then I would not try to pull the domain from the computer name.. Otherwise there can be false/incorrect information..

Hope that makes sense, if not let me know :)

janniten replied:
@neu5ron you are correct. Thank you for your feedback.
If the computer is not joined or member of the domain I cannot fetch the domain name from the computer name.
In the case of the audit events like 1100, there is no other field in the event from where the domain name can be obtained; but to keep consistency and correctness I'll remove the addComputerData function

janniten replied:
@neu5ron @leehinman I have removed the addComputerData function.
Please let me know if I have to modify something else
Thank you!

leehinman replied:
@janniten I think the PR needs to be rebased so there aren't merge conflicts and the CI tests will run.

janniten replied:
@leehinman . Done!

leehinman replied:
@janniten sorry, I forgot to say "rebased off a more recent master" . It looks like this PR is still based on 9b48f2bbb2 from Wed Dec 18 11:43:56 2019 +0100.  Unfortunately 586eac17e8 on Fri Dec 20 13:54:27 2019 -0600 makes changes to the golden files in the same place this PR does, so they conflict.

janniten replied:
@leehinman Let me know I know is ok. Thank you!

leehinman commented:
jenkins, test this

andrewkroh commented:
jenkins, test this

andrewkroh commented on x-pack/winlogbeat/module/security/config/winlogbeat-security.js:
```diff
@@ -1172,6 +1264,36 @@ var security = (function () {
+
+    var addRelatedUser= function(evt,user) {
+        var related_user = evt.Get("related.user");
+        if (!related_user) {
```
From and Elasticsearch document standpoint this is not necessary. You can just call `evt.AppendTo` in both cases and it will work. Whether the value is a string or an array or strings it's going to be handled the same way in ES.

janniten replied:
I have modified the code in order to use directly the evt.AppentTo(related.user,user_to_add), less functions calls, less code, more efficiency :)
Thank you for the suggestion. I'll have this in mind for futures PRs

janniten replied:
@andrewkroh some checks fail regarding to the documentation update.
I have added the events ID in \beats\x-pack\winlogbeat\module\security\_meta\docs.asciidoc.
Shall I do something more?

![image](https://user-images.githubusercontent.com/33020901/73177284-e543e100-410e-11ea-995d-f3ce4d548216.png)

janniten commented:
> We need to update the list of event IDs in the module's asciidoc file. Can you do that in this PR please?
> Thanks for your contribution.

@andrewkroh I have aldready added the events ID. I have also added the ones in https://github.com/elastic/beats/pull/14299 that alse were missing.

elasticmachine commented:
Pinging @elastic/siem (Team:SIEM)

andrewkroh commented:
jenkins, test this

andrewkroh commented:
jenkins, test this

andrewkroh commented:
jenkins, test this

andrewkroh commented:
🍾 Thank you @janniten for your contributions to Winlogbeat and ECS!
